examples/evaluation/evaluate_with_custom_metrics_comprehensive.py

```diff
@@ -0,0 +1,166 @@
+    """
+    return {
+        "squared_diff_plus_one": np.sum(np.abs(eval_df["prediction"] - eval_df["target"] + 1) ** 2),
+        "sum_on_label_divided_by_two": builtin_metrics["sum_on_label"] / 2,
+    }
+
+
+def file_artifacts_fn(eval_df, builtin_metrics):
+    """
+    This example shows how you can return file paths as representation
+    of the produced artifacts. For a full list of supported file extensions
+    refer to https://mlflow.org/docs/latest/python_api/mlflow.html#mlflow.evaluate
+    :param eval_df:
+    :param builtin_metrics:
+    :return:
+    """
+    example_np_arr = np.array([1, 2, 3])
+    np.save(os.path.join(tmp_dir, "example.npy"), example_np_arr, allow_pickle=False)
+
+    example_df = pd.DataFrame({"test": [2.2, 3.1], "test2": [3, 2]})
+    example_df.to_csv(os.path.join(tmp_dir, "example.csv"), index=False)
+    example_df.to_parquet(os.path.join(tmp_dir, "example.parquet"))
+
+    example_json = {"hello": "there", "test_list": [0.1, 0.3, 4]}
+    example_json.update(builtin_metrics)
+    with open(os.path.join(tmp_dir, "example.json"), "w") as f:
+        json.dump(example_json, f)
+
+    plt.scatter(eval_df["prediction"], eval_df["target"])
+    plt.xlabel("Targets")
+    plt.ylabel("Predictions")
+    plt.title("Targets vs. Predictions")
+    plt.savefig(os.path.join(tmp_dir, "example.png"))
+    plt.savefig(os.path.join(tmp_dir, "example.jpeg"))
+
+    with open(os.path.join(tmp_dir, "example.txt"), "w") as f:
+        f.write("hello world!")
+
+    return {}, {
+        "example_np_arr_from_npy_file": os.path.join(tmp_dir, "example.npy"),
+        "example_df_from_csv_file": os.path.join(tmp_dir, "example.csv"),
+        "example_df_from_parquet_file": os.path.join(tmp_dir, "example.parquet"),
+        "example_dict_from_json_file": os.path.join(tmp_dir, "example.json"),
+        "example_image_from_png_file": os.path.join(tmp_dir, "example.png"),
+        "example_image_from_jpeg_file": os.path.join(tmp_dir, "example.jpeg"),
+        "example_string_from_txt_file": os.path.join(tmp_dir, "example.txt"),
+    }
+
+
+class ExampleClass:
+    def __init__(self, x):
+        self.x = x
+
+
+def object_artifacts_fn(eval_df, builtin_metrics):
+    """
+    This example shows how you can return python objects as artifacts
+    without the need to save them to file system.
+    """
+    example_np_arr = np.array([1, 2, 3])
+    example_df = pd.DataFrame({"test": [2.2, 3.1], "test2": [3, 2]})
+    example_dict = {"hello": "there", "test_list": [0.1, 0.3, 4]}
+    example_dict.update(builtin_metrics)
+    example_dict_2 = '{"a": 3, "b": [1, 2, 3]}'
+    example_image = plt.figure()
+    plt.scatter(eval_df["prediction"], eval_df["target"])
+    plt.xlabel("Targets")
+    plt.ylabel("Predictions")
+    plt.title("Targets vs. Predictions")
+    example_custom_class = ExampleClass(10)
+
+    # In rare cases, you might already have constructed a EvaluationArtifact-like object
```

Comment:
Why is this rare?

Reply:
The common use case I can see for this is letting the user return a custom evaluation artifact type that isn't already supported. Do we have a recommended workflow for that?

Reply:
Will include an example for this in the follow up PR discussed [here](https://github.com/mlflow/mlflow/pull/5405#issuecomment-1062114812)